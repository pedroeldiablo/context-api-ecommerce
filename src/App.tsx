import React from 'react';
import { Switch, Route, Redirect } from 'react-router-dom';

import './App.css';

import { HomePage } from './pages/homepage/homepage.component';
import { ShopPage } from './pages/shop/shop.component';
import { SignInAndSignUpPage} from './pages/sign-in-and-sign-up/sign-in-and-sign-up.component';
import { CheckoutPage } from './pages/checkout/checkout.component';
import { PageNotFoundPage } from './pages/404/404.component'

import { Header } from './components/header/header.component';

import { auth, createUserProfileDocument } from './firebase/firebase.utils';
import { Unsubscribe } from 'redux';
import {CurrentUserContext} from './contexts/current-user/current-user.context';

interface AppProps {
  props: any
}

interface AppState {
  currentUser: null | any
  
}

export class App extends React.Component<AppProps, AppState> {
  constructor(props: any) {
    super(props);

    this.state = {
      currentUser: null,
    }

  }
  // unsubscribeFromAuth: null | Unsubscribe = null;

  unsubscribeFromAuth: null | Unsubscribe  = null;

  componentDidMount() {
    this.unsubscribeFromAuth = auth.onAuthStateChanged(async userAuth => {
      if (userAuth) {
        const userRef = await createUserProfileDocument(userAuth);

        userRef?.onSnapshot(snapShot => {
          this.setState({currentUser: {
            id: snapShot.id,
            ...snapShot.data()
          }});
        });
      }

      this.setState({ currentUser: userAuth});
    });
  }

  componentWillUnmount() {
    if(this.unsubscribeFromAuth){
      this.unsubscribeFromAuth();
    }
   
  }

  render() {
    return (
      <div>
        <CurrentUserContext.Provider value={this.state.currentUser}>
          <Header></Header>
        </CurrentUserContext.Provider>
        <Switch>
          <Route exact path='/' component={HomePage} />
          <Route path='/shop' component={ShopPage} />
          <Route exact path='/checkout' component={CheckoutPage} />
          <Route
            exact
            path='/signin'
            render={() =>
              this.state.currentUser ? (
                <Redirect to='/' />
              ) : (
                <SignInAndSignUpPage />
              )
            }
          />
          <Route component={PageNotFoundPage} />
        </Switch>
      </div>
    );
  }
}
