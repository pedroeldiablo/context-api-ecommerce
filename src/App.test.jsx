import React from 'react';
import ReactDOM from 'react-dom';
import { App } from './App';
import renderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom';
import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
// import { createMemoryHistory } from 'history'
import '@testing-library/jest-dom/extend-expect'
import {renderWithRouter} from './test/utils/renderWithRouter'


it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render((
   <BrowserRouter>
    <App />
   </BrowserRouter>), div);
 });

describe('Test suits for MyComponentWithLink', () => {
  it('should match with snapshot', () => {
  const tree = renderer
    .create(
      <BrowserRouter>
        <App/>
      </BrowserRouter>
    )
    .toJSON();
   expect(tree).toMatchSnapshot();
   });
  });

  test('rendering/navigating', () => {
    //   const history = createMemoryHistory()
      render(
         <BrowserRouter>
         <App/>
       </BrowserRouter>
      )
      // verify page content for expected route
      // often you'd use a data-testid or role query, but this is also possible
      expect(screen.getByText(/CONTACT/i)).toBeInTheDocument()
    
      const leftClick = { button: 0 }
      userEvent.click(screen.getByText(/^MENS$/i), leftClick)
    
      // check that the content changed to the new page
      expect(screen.getByText(/^Camo Down Vest$/i)).toBeInTheDocument()
    })

    test('landing on a bad page', () => {
      renderWithRouter(<App />, { route: '/gobbledygook' })
    
      expect(screen.getByText(/^Oops seems there's nothing here$/i)).toBeInTheDocument()
    })
