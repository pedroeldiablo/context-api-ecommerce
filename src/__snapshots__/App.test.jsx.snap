// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test suits for MyComponentWithLink should match with snapshot 1`] = `
<div>
  <div
    className="header"
  >
    <a
      className="logo-container"
      href="/"
      onClick={[Function]}
    >
      <svg
        className="logo"
      >
        crown.svg
      </svg>
    </a>
    <div
      className="options"
    >
      <a
        className="option"
        href="/shop"
        onClick={[Function]}
      >
        SHOP
      </a>
      <a
        className="option"
        href="/contact"
        onClick={[Function]}
      >
        CONTACT
      </a>
      <a
        className="option"
        href="/signin"
        onClick={[Function]}
      >
        SIGN IN
      </a>
      <div
        className="cart-icon"
        onClick={[Function]}
      >
        <svg
          className="shopping-icon"
        >
          shopping-bag.svg
        </svg>
        <span
          className="item-count"
        >
          0
        </span>
      </div>
    </div>
  </div>
  <div
    className="homepage"
  >
    <div
      className="directory-menu"
    >
      <div
        className="undefined menu-item"
        onClick={[Function]}
      >
        <div
          className="background-image"
          style={
            Object {
              "backgroundImage": "url(https://i.ibb.co/cvpntL1/hats.png)",
            }
          }
        />
        <div
          className="content"
        >
          <h1
            className="title"
          >
            HATS
          </h1>
          <span
            className="subtitle"
          >
            SHOP NOW
          </span>
        </div>
      </div>
      <div
        className="undefined menu-item"
        onClick={[Function]}
      >
        <div
          className="background-image"
          style={
            Object {
              "backgroundImage": "url(https://i.ibb.co/px2tCc3/jackets.png)",
            }
          }
        />
        <div
          className="content"
        >
          <h1
            className="title"
          >
            JACKETS
          </h1>
          <span
            className="subtitle"
          >
            SHOP NOW
          </span>
        </div>
      </div>
      <div
        className="undefined menu-item"
        onClick={[Function]}
      >
        <div
          className="background-image"
          style={
            Object {
              "backgroundImage": "url(https://i.ibb.co/0jqHpnp/sneakers.png)",
            }
          }
        />
        <div
          className="content"
        >
          <h1
            className="title"
          >
            SNEAKERS
          </h1>
          <span
            className="subtitle"
          >
            SHOP NOW
          </span>
        </div>
      </div>
      <div
        className="large menu-item"
        onClick={[Function]}
      >
        <div
          className="background-image"
          style={
            Object {
              "backgroundImage": "url(https://i.ibb.co/GCCdy8t/womens.png)",
            }
          }
        />
        <div
          className="content"
        >
          <h1
            className="title"
          >
            WOMENS
          </h1>
          <span
            className="subtitle"
          >
            SHOP NOW
          </span>
        </div>
      </div>
      <div
        className="large menu-item"
        onClick={[Function]}
      >
        <div
          className="background-image"
          style={
            Object {
              "backgroundImage": "url(https://i.ibb.co/R70vBrQ/men.png)",
            }
          }
        />
        <div
          className="content"
        >
          <h1
            className="title"
          >
            MENS
          </h1>
          <span
            className="subtitle"
          >
            SHOP NOW
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
